File: dataforge-control/src/main/java/hep/dataforge/control/DeviceManager.java
package hep.dataforge.control;

import hep.dataforge.connections.Connection;
import hep.dataforge.context.*;
import hep.dataforge.control.devices.Device;
import hep.dataforge.control.devices.DeviceFactory;
import hep.dataforge.control.devices.DeviceHub;
import hep.dataforge.exceptions.ControlException;
import hep.dataforge.exceptions.EnvelopeTargetNotFoundException;
import hep.dataforge.io.envelopes.Envelope;
import hep.dataforge.io.messages.Dispatcher;
import hep.dataforge.io.messages.Responder;
import hep.dataforge.meta.Meta;
import hep.dataforge.names.Name;
import org.jetbrains.annotations.NotNull;

import java.util.HashMap;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import java.util.stream.Stream;

/**
 * A plugin for creating and using different devices
 * Created by darksnake on 11-Oct-16.
 */
@PluginDef(name = "devices", info = "Management plugin for devices an their interaction")
public class DeviceManager extends BasicPlugin implements Dispatcher, DeviceHub {

    /**
     * Registered devices
     */
    private Map<Name, Device> devices = new HashMap<>();


    public void add(Device device) {
        Name name = Name.ofSingle(device.getName());
        if (devices.containsKey(name)) {
            getLogger().warn("Replacing existing device in device manager!");
            remove(name);
        }
        devices.put(name, device);
    }

    public void remove(Name name) {
        Optional.ofNullable(this.devices.remove(name)).ifPresent(it -> {
            try {
                it.shutdown();
            } catch (ControlException e) {
                getLogger().error("Failed to stop the device: " + it.getName(), e);
            }
        });
    }


    public Device buildDevice(Meta deviceMeta) {
        DeviceFactory factory = getContext()
                .optService(DeviceFactory.class, f -> Objects.equals(f.getType(), ControlUtils.getDeviceType(deviceMeta)))
                .orElseThrow(() -> new RuntimeException("Can't find factory for given device type"));
        Device device = factory.build(getContext(), deviceMeta);

        deviceMeta.getMetaList("connection").forEach(connectionMeta -> {
            device.getConnectionHelper().connect(getContext(), connectionMeta);
        });

        add(device);
        return device;
    }

    @NotNull
    @Override
    public Responder getResponder(@NotNull Meta targetInfo) throws EnvelopeTargetNotFoundException {
        throw new UnsupportedOperationException();
    }

    @NotNull
    @Override
    public Responder getResponder(@NotNull Envelope envelope) {
        throw new UnsupportedOperationException();
    }

    @Override
    public Optional<Device> optDevice(Name name) {
        if (name.isEmpty()) {
            throw new IllegalArgumentException("Can't provide a device with zero name");
        } else if (name.getLength() == 1) {
            return Optional.ofNullable(devices.get(name));
        } else {
            return Optional.ofNullable(devices.get(name.getFirst())).flatMap(hub -> {
                if (hub instanceof DeviceHub) {
                    return ((DeviceHub) hub).optDevice(name.cutFirst());
                } else {
                    return Optional.empty();
                }
            });
        }
    }

    @Override
    public Stream<Name> getDeviceNames() {
        return devices.entrySet().stream().flatMap(entry -> {
            if (entry.getValue() instanceof DeviceHub) {
                return ((DeviceHub) entry.getValue()).getDeviceNames().map(it -> entry.getKey().append(it));
            } else {
                return Stream.of(entry.getKey());
            }
        });
    }

    /**
     * Get the stream of top level devices
     *
     * @return
     */
    public Stream<Device> getDevices() {
        return devices.values().stream();
    }

    @Override
    public void detach() {
        devices.values().forEach(it -> {
            try {
                it.shutdown();
            } catch (ControlException e) {
                getLogger().error("Failed to stop the device: " + it.getName(), e);
            }
        });
        super.detach();
    }

    @Override
    public void connectAll(Connection connection, String... roles) {
        this.devices.values().forEach(device -> device.connect(connection, roles));
    }

    @Override
    public void connectAll(Context context, Meta meta) {
        this.devices.values().forEach(device -> device.getConnectionHelper().connect(context, meta));
    }

    public static class Factory extends PluginFactory {
        @NotNull
        @Override
        public Plugin build(@NotNull Meta meta) {
            return new DeviceManager();
        }

        @NotNull
        @Override
        public Class<? extends Plugin> getType() {
            return DeviceManager.class;
        }
    }
}

File: dataforge-control/src/main/kotlin/hep/dataforge/control/DeviceManager.kt
/*
 * Copyright  2018 Alexander Nozik.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

package hep.dataforge.control

import hep.dataforge.connections.Connection
import hep.dataforge.context.*
import hep.dataforge.control.devices.Device
import hep.dataforge.control.devices.DeviceFactory
import hep.dataforge.control.devices.DeviceHub
import hep.dataforge.exceptions.ControlException
import hep.dataforge.exceptions.EnvelopeTargetNotFoundException
import hep.dataforge.io.envelopes.Envelope
import hep.dataforge.io.messages.Dispatcher
import hep.dataforge.io.messages.Responder
import hep.dataforge.meta.Meta
import hep.dataforge.names.Name
import java.util.*
import java.util.stream.Stream

/**
 * A plugin for creating and using different devices
 * Created by darksnake on 11-Oct-16.
 */
@PluginDef(name = "devices", info = "Management plugin for devices an their interaction")
class DeviceManager : BasicPlugin(), Dispatcher, DeviceHub {

    /**
     * Registered devices
     */
    private val devices = HashMap<Name, Device>()

    override val deviceNames: Stream<Name>
        get() = devices.entries.stream().flatMap { entry ->
            if (entry.value is DeviceHub) {
                (entry.value as DeviceHub).deviceNames.map { it -> entry.key.append(it) }
            } else {
                Stream.of<Name>(entry.key)
            }
        }


    fun add(device: Device) {
        val name = Name.ofSingle(device.name)
        if (devices.containsKey(name)) {
            logger.warn("Replacing existing device in device manager!")
            remove(name)
        }
        devices[name] = device
    }

    fun remove(name: Name) {
        Optional.ofNullable(this.devices.remove(name)).ifPresent { it ->
            try {
                it.shutdown()
            } catch (e: ControlException) {
                logger.error("Failed to stop the device: " + it.name, e)
            }
        }
    }


    fun buildDevice(deviceMeta: Meta): Device {
        val factory = context
                .optService(DeviceFactory::class.java) { it.type == ControlUtils.getDeviceType(deviceMeta) }
                .orElseThrow { RuntimeException("Can't find factory for given device type") }
        val device = factory.build(context, deviceMeta)

        deviceMeta.getMetaList("connection").forEach { connectionMeta -> device.connectionHelper.connect(context, connectionMeta) }

        add(device)
        return device
    }

    @Throws(EnvelopeTargetNotFoundException::class)
    override fun getResponder(targetInfo: Meta): Responder {
        throw UnsupportedOperationException()
    }

    override fun getResponder(envelope: Envelope): Responder {
        throw UnsupportedOperationException()
    }

    override fun optDevice(name: Name): Optional<Device> {
        return when {
            name.isEmpty -> throw IllegalArgumentException("Can't provide a device with zero name")
            name.length == 1 -> Optional.ofNullable(devices[name])
            else -> Optional.ofNullable(devices[name.first]).flatMap { hub ->
                if (hub is DeviceHub) {
                    (hub as DeviceHub).optDevice(name.cutFirst())
                } else {
                    Optional.empty()
                }
            }
        }
    }

    /**
     * Get the stream of top level devices
     *
     * @return
     */
    fun getDevices(): Stream<Device> {
        return devices.values.stream()
    }

    override fun detach() {
        devices.values.forEach { it ->
            try {
                it.shutdown()
            } catch (e: ControlException) {
                logger.error("Failed to stop the device: " + it.name, e)
            }
        }
        super.detach()
    }

    override fun connectAll(connection: Connection, vararg roles: String) {
        this.devices.values.forEach { device -> device.connect(connection, *roles) }
    }

    override fun connectAll(context: Context, meta: Meta) {
        this.devices.values.forEach { device -> device.connectionHelper.connect(context, meta) }
    }

    class Factory : PluginFactory() {

        override val type: Class<out Plugin>
            get() = DeviceManager::class.java

        override fun build(meta: Meta): Plugin {
            return DeviceManager()
        }
    }
}
